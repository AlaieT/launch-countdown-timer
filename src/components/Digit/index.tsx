import React, { useEffect, useState } from "react";

import type { DigitProps } from "../../types";

import styles from "../../styles/components/digit.module.scss";
import classNames from "../../utils/classNames";

function Digit({ label, digit }: DigitProps) {
  const [lastDigit, setLastDigit] = useState(digit);
  const [isAnimated, setIsAnimated] = useState(false);

  useEffect(() => {
    if (lastDigit !== digit) {
      setIsAnimated((item) => !item);

      const timeout = setTimeout(() => {
        setIsAnimated((item) => !item);
        setLastDigit(digit);
        clearTimeout(timeout);
      }, 500);

      return () => clearTimeout(timeout);
    }

    return () => undefined;
  }, [digit, lastDigit]);

  return (
    <div className={styles.digit}>
      <div className={styles.box}>
        <div className={styles.top}>
          <p>
            {digit < 10 && "0"}
            {digit}
          </p>
        </div>
        <div
          className={classNames({
            [styles.middle]: true,
            [styles.animated]: isAnimated
          })}
        >
          <div className={styles.front}>
            <p>
              {lastDigit < 10 && "0"}
              {lastDigit}
            </p>
          </div>
          <div className={styles.back}>
            <p>
              {digit < 10 && "0"}
              {digit}
            </p>
          </div>
        </div>
        <div className={styles.bottom}>
          <p>
            {lastDigit < 10 && "0"}
            {lastDigit}
          </p>
        </div>
      </div>
      <div className={styles.label}>
        <p>{label}</p>
      </div>
    </div>
  );
}

export default Digit;
