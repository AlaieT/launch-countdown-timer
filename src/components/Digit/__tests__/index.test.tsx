import React from "react";
import { render } from "@testing-library/react";

import Digit from "../index";

describe("Digit testing...", () => {
  it("Snapshot testing", () => {
    const { asFragment } = render(<Digit label="day" digit={24} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
