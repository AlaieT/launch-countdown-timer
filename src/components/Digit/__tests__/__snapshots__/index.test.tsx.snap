// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Digit testing... Snapshot testing 1`] = `
<DocumentFragment>
  <div
    class="digit"
  >
    <div>
      <div>
        Real:24
      </div>
      <div>
        Prev:24
      </div>
    </div>
    <p>
      day
    </p>
  </div>
</DocumentFragment>
`;
